reconcile/aus/advanced_upgrade_service.py

```diff
@@ -327,6 +410,92 @@ def _build_policy_from_labels(labels: LabelContainer) -> ClusterUpgradePolicyV1:
+    A list of OCM organization IDs to inherit version data from. These organization also need
+    to publish their version data via the `publish-version-data` label to the inheriting version.
+    Version data publishing/inheritance can also be defined between OCM environments.
+    """
+
+    publish_version_data: Optional[CSV] = Field(
+        alias=aus_label_key("version-data.publish")
+    )
+    """
+    A list of OCM organization IDs to publish version data to. These organization also need
+    to explicitely inherit version data via the `inherit-version-data` label from this organization.
+    Version data publishing/inheritance can also be defined between OCM environments.
+    """
+
+
+@dataclass(frozen=True, eq=True)
+class OrgRef:
+    org_id: str
+    env_name: str
+
+
+class VersionDataInheritance(BaseModel):
+    org_id: str
+    inherit_from_orgs: set[OrgRef]
+    unverified_inheritance_from_orgs: set[OrgRef]
+
+
+def build_version_data_inheritance_network(
+    labels_per_org: dict[OrgRef, LabelContainer]
+) -> dict[OrgRef, VersionDataInheritance]:
+    """
+    Validates publish/inherit relationships between OCM organizations and environments from the
+    provided label containers.
+
+    This function returns a dictionary of OCM organizations and their version data
+    inheritance relationships.
+    """
+    label_set_per_org = {
+        org_ref: build_labelset(labels, VersionDataInheritanceLabelSet)
+        for org_ref, labels in labels_per_org.items()
+    }
+    org_ref_lookup = {org_ref.org_id: org_ref for org_ref in labels_per_org}
+
+    verified_inheritance_network: dict[OrgRef, VersionDataInheritance] = {}
+    for org_ref, label_set in label_set_per_org.items():
+        if not label_set.inherit_version_data:
+            continue
+
+        verified_inheritance_network[org_ref] = VersionDataInheritance(
+            org_id=org_ref.org_id,
+            inherit_from_orgs={
+                org_ref_lookup[source_org_id]
+                for source_org_id in label_set.inherit_version_data
+                if source_org_id in org_ref_lookup
+                and org_ref.org_id
+                in (
+                    label_set_per_org[
+                        org_ref_lookup[source_org_id]
+                    ].publish_version_data
+                    or []
+                )
+            },
+            unverified_inheritance_from_orgs={
+                org_ref_lookup.get(
+                    source_org_id, OrgRef(org_id=source_org_id, env_name="unknown")
+                )
+                for source_org_id in label_set.inherit_version_data
+                if source_org_id not in org_ref_lookup
+                or org_ref.org_id
+                not in (
+                    label_set_per_org[
+                        org_ref_lookup[source_org_id]
+                    ].publish_version_data
+                    or []
+                )
+            },
+        )
+
+    return verified_inheritance_network
```

Comment:
`label_set.inherit_versino_data` can either be `inherit_from_orgs`, or `unverified_inheritance_from_orgs`, right? Maybe clearer this way

```python
def _build_version_data_inheritance(
    org_ref,
    label_set,
    org_ref_lookup,
    label_set_per_org,
):
    inherit_from_orgs_org_ids = {
        source_org_id
        for source_org_id in label_set.inherit_version_data
        if source_org_id in org_ref_lookup
        and org_ref.org_id
        in (label_set_per_org[org_ref_lookup[source_org_id]].publish_version_data or [])
    }
    unverified_inheritance_from_orgs_org_ids = (
        set(label_set.inherit_version_data) - inherit_from_orgs_org_ids
    )

    return VersionDataInheritance(
        org_id=org_ref.org_id,
        inherit_from_orgs={
            org_ref_lookup[source_org_id] for source_org_id in inherit_from_orgs_org_ids
        },
        unverified_inheritance_from_orgs={
            org_ref_lookup.get(
                source_org_id, OrgRef(org_id=source_org_id, env_name="unknown")
            )
            for source_org_id in unverified_inheritance_from_orgs_org_ids
        },
    )
```

Then it can be used like

```python
    return {
        org_ref: _build_version_data_inheritance(
            org_ref,
            label_set,
            org_ref_lookup,
            label_set_per_org,
        )
        for org_ref, label_set in label_set_per_org.items()
        if label_set.inherit_version_data
    }
```

Reply:
i incorporated your code. thank you ❤️